dm/syncer/ddl.go

```diff
@@ -18,18 +18,912 @@ import (
+			zap.Stringer("start location", startLocation),
+			log.WrapStringerField("end location", endLocation))
+		err = ddl.safeMode.DescForTable(qec.tctx, ddlInfo.targetTables[0]) // try disable safe-mode after sharding group synced
+		if err != nil {
+			return err
+		}
+		// maybe multi-groups' sharding DDL synced in this for-loop (one query-event, multi tables)
+		if cap(*qec.shardingReSyncCh) < len(needHandleDDLs) {
+			*qec.shardingReSyncCh = make(chan *ShardingReSync, len(needHandleDDLs))
+		}
+		firstEndLocation := group.FirstEndPosUnresolved()
+		if firstEndLocation == nil {
+			return terror.ErrSyncerUnitFirstEndPosNotFound.Generate(sourceTableID)
+		}
+
+		allResolved, err2 := ddl.sgk.ResolveShardingDDL(ddlInfo.targetTables[0])
+		if err2 != nil {
+			return err2
+		}
+		*qec.shardingReSyncCh <- &ShardingReSync{
+			currLocation:   *firstEndLocation,
+			latestLocation: endLocation,
+			targetTable:    ddlInfo.targetTables[0],
+			allResolved:    allResolved,
+		}
+
+		// Don't send new DDLInfo to dm-master until all local sql jobs finished
+		// since jobWg is flushed by flushJobs before, we don't wait here any more
+
+		// NOTE: if we need singleton Syncer (without dm-master) to support sharding DDL sync
+		// we should add another config item to differ, and do not save DDLInfo, and not wait for ddlExecInfo
+
+		// construct & send shard DDL info into etcd, DM-master will handle it.
+		shardInfo := ddl.pessimist.ConstructInfo(ddlInfo.targetTables[0].Schema, ddlInfo.targetTables[0].Name, needHandleDDLs)
+		rev, err2 := ddl.pessimist.PutInfo(qec.tctx.Ctx, shardInfo)
+		if err2 != nil {
+			return err2
+		}
+		ddl.metricsProxies.Metrics.ShardLockResolving.Set(1) // block and wait DDL lock to be synced
+		ddl.logger.Info("putted shard DDL info", zap.Stringer("info", shardInfo), zap.Int64("revision", rev))
+
+		shardOp, err2 := ddl.pessimist.GetOperation(qec.tctx.Ctx, shardInfo, rev+1)
+		ddl.metricsProxies.Metrics.ShardLockResolving.Set(0)
+		if err2 != nil {
+			return err2
+		}
+
+		if shardOp.Exec {
+			failpoint.Inject("ShardSyncedExecutionExit", func() {
+				ddl.logger.Warn("exit triggered", zap.String("failpoint", "ShardSyncedExecutionExit"))
+				//nolint:errcheck
+				ddl.flushCheckPoints()
+				utils.OsExit(1)
+			})
+			failpoint.Inject("SequenceShardSyncedExecutionExit", func() {
+				group := ddl.sgk.Group(ddlInfo.targetTables[0])
+				if group != nil {
+					// exit in the first round sequence sharding DDL only
+					if group.meta.ActiveIdx() == 1 {
+						ddl.logger.Warn("exit triggered", zap.String("failpoint", "SequenceShardSyncedExecutionExit"))
+						//nolint:errcheck
+						ddl.flushCheckPoints()
+						utils.OsExit(1)
+					}
+				}
+			})
+
+			ddl.logger.Info("execute DDL job",
+				zap.String("event", "query"),
+				zap.Stringer("queryEventContext", qec),
+				zap.String("sourceTableID", sourceTableID),
+				zap.Stringer("operation", shardOp))
+		} else {
+			ddl.logger.Info("ignore DDL job",
```

Comment:
seems this is the same with old logic, but why we print "ignore" and continue building the DDL job 😂